onnxruntime/core/mlas/lib/tanh.cpp

```diff
@@ -83,69 +73,48 @@ Return Value:
 
 --*/
 {
-    while (N >= 4) {
-
-        MLAS_FLOAT32X4 Value = MlasLoadFloat32x4(Input);
-
-        Value = MlasMaximumFloat32x4(MlasBroadcastFloat32x4(MlasTanhConstants.LowerRange), Value);
-        Value = MlasMinimumFloat32x4(MlasBroadcastFloat32x4(MlasTanhConstants.UpperRange), Value);
-
-        MLAS_FLOAT32X4 ValueSquared = MlasMultiplyFloat32x4(Value, Value);
-
-        MLAS_FLOAT32X4 p;
-        p = MlasMultiplyAddFloat32x4(ValueSquared, MlasBroadcastFloat32x4(MlasTanhConstants.alpha_13),
-            MlasBroadcastFloat32x4(MlasTanhConstants.alpha_11));
-        p = MlasMultiplyAddFloat32x4(p, ValueSquared, MlasBroadcastFloat32x4(MlasTanhConstants.alpha_9));
-        p = MlasMultiplyAddFloat32x4(p, ValueSquared, MlasBroadcastFloat32x4(MlasTanhConstants.alpha_7));
-        p = MlasMultiplyAddFloat32x4(p, ValueSquared, MlasBroadcastFloat32x4(MlasTanhConstants.alpha_5));
-        p = MlasMultiplyAddFloat32x4(p, ValueSquared, MlasBroadcastFloat32x4(MlasTanhConstants.alpha_3));
-        p = MlasMultiplyAddFloat32x4(p, ValueSquared, MlasBroadcastFloat32x4(MlasTanhConstants.alpha_1));
-        p = MlasMultiplyFloat32x4(p, Value);
-
-        MLAS_FLOAT32X4 q;
-        q = MlasMultiplyAddFloat32x4(ValueSquared, MlasBroadcastFloat32x4(MlasTanhConstants.beta_6),
-            MlasBroadcastFloat32x4(MlasTanhConstants.beta_4));
-        q = MlasMultiplyAddFloat32x4(q, ValueSquared, MlasBroadcastFloat32x4(MlasTanhConstants.beta_2));
-        q = MlasMultiplyAddFloat32x4(q, ValueSquared, MlasBroadcastFloat32x4(MlasTanhConstants.beta_0));
-
-        MlasStoreFloat32x4(Output, MlasDivideFloat32x4(p, q));
+    const MLAS_FLOAT32X4 nc0 = MlasBroadcastFloat32x4(reinterpret_cast<const float*>(&MlasTanhConstants._nc0));
+    const MLAS_FLOAT32X4 nc1 = MlasBroadcastFloat32x4(reinterpret_cast<const float*>(&MlasTanhConstants._nc1));
+    const MLAS_FLOAT32X4 nc2 = MlasBroadcastFloat32x4(reinterpret_cast<const float*>(&MlasTanhConstants._nc2));
+    const MLAS_FLOAT32X4 dc0 = MlasBroadcastFloat32x4(reinterpret_cast<const float*>(&MlasTanhConstants._dc0));
+    const MLAS_FLOAT32X4 dc1 = MlasBroadcastFloat32x4(reinterpret_cast<const float*>(&MlasTanhConstants._dc1));
+    const MLAS_FLOAT32X4 dc2 = MlasBroadcastFloat32x4(reinterpret_cast<const float*>(&MlasTanhConstants._dc2));
+    const MLAS_FLOAT32X4 ub = MlasBroadcastFloat32x4(reinterpret_cast<const float*>(&MlasTanhConstants._ubound));
+    const MLAS_FLOAT32X4 absmask = MlasBroadcastFloat32x4(reinterpret_cast<const float*>(&MlasTanhConstants._absmask));
+    MLAS_FLOAT32X4 Val;
+
+    size_t count = 0;
+    while (count < N) {
+        if (N - count >= 4) {
```

Comment:
there needs a check for each iteration with this change. If N is large, the previous version can save a significant amount of instructions

Reply:
For large values of N, the CPU branch predictor should be able to branch predict this branch pretty easily. It will only miss at the very last iteration for the tail but when N is large, one single branch miss should hardly matter in terms of performance. It does bring the benefit of processing the entire array contained in a single loop.